core/thread/auth.go

```diff
@@ -0,0 +1,54 @@
+package thread
+
+import (
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+)
+
+// Auth provides a signed identity, which is needed for thread ACL checks.
+type Auth struct {
+	sk  crypto.PrivKey
+	pk  crypto.PubKey
+	msg []byte
+	sig []byte
+}
+
+// NewAuthFromPrivKey returns auth from a private key.
+// The auth must be signed before use.
+func NewAuthFromPrivKey(privKey crypto.PrivKey) *Auth {
+	return &Auth{sk: privKey}
+}
+
+// NewAuthFromSignature returns auth from a public key and signature.
+func NewAuthFromSignature(pubKey crypto.PubKey, sig []byte) *Auth {
+	return &Auth{pk: pubKey, sig: sig}
+}
+
+func (a *Auth) Sign(msg []byte) (sig []byte, pk crypto.PubKey, err error) {
+	a.msg = msg
+	if a.sk != nil {
+		sig, err := a.sk.Sign(msg)
+		if err != nil {
+			return nil, nil, err
+		}
+		return sig, a.sk.GetPublic(), nil
+	}
+	return a.sig, a.pk, nil
+}
```

Comment:
If `Auth` was created with `NewAuthFromSignature`, then `msg` is ignored.
That seems quite weird, and a bit dangerous... a `Sign(msg)` method that ignores what is being signed, but returns a signature anyway.

Reply:
This should make sense when you see it in action... the API does this in order to set the message to be validated. I guess it could be an additional API.

Reply:
Uhm, yeah. I'd say a different API to do that sounds better, since in that case, it isn't signing anything.

I still have an itch on this logic of having to save a signature, which doesn't correspond always with the corresponding message. Still, something doesn't tick for me.

Reply:
I don't love it either... the problem is that I'm trying use the same object for,
- A user that has a private key they want kept private, but they don't know how to sign the message since the message creation is done internally... hence the Sign method being called from within the client.
- The API that receives the signed message and pubkey, and then must also construct a `thread.Auth` message for use with the _internal_ thread net API... hence the need to set the message.

I think ultimately, the message being signed might have to change from being a protobuf message to being some custom combination of the method arguments for each API call. In this way, the server could just pass through the "auth" without have to set the message.

Reply:
Hmm, yeah I think that's better... more tedious, but it would avoid this problem. And another problem that is more subtle... a pure-Go app that doesn't use the gRPC APIs could totally cheat the current setup.

Reply:
I agree that using the same struct for both things is what's causing the problem.
For one use case, it seems that the API should allow receiving some `SignedMessage` (which may be a plain struct without methods). And send that to the internal API to deal with that.

Forcing to use the same struct, makes this _fake signings_.

Reply:
That might be an indication that the API should always receive `SignedMessages` and let signing be a problem outside the API. If comes through gRPC, was done by the client. If not, in some other place.

_sharing thoughts, not sure covers all the cases_

Reply:
Yep, if the non-gRPC clients _know how to sign_, then we can do that... that mean the message to be signed is some combo of the method args, not the nicely signable proto messages.